homeassistant/components/vilfo/__init__.py

```diff
@@ -0,0 +1,120 @@
+from vilfo.exceptions import VilfoException
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_ACCESS_TOKEN, CONF_HOST
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+from homeassistant.util import Throttle, dt as dt_util
+
+from .const import ATTR_DEFAULT_HOST, DOMAIN
+
+PLATFORMS = ["sensor"]
+
+DEFAULT_SCAN_INTERVAL = timedelta(seconds=30)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistantType, config: ConfigType):
+    """Set up the Vilfo Router component."""
+    hass.data.setdefault(DOMAIN, {})
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Vilfo Router from a config entry."""
+    host = entry.data[CONF_HOST]
+    access_token = entry.data[CONF_ACCESS_TOKEN]
+
+    vilfo_router = VilfoRouterData(host, access_token)
+
+    await vilfo_router.async_update()
+
+    if not vilfo_router.available:
+        raise ConfigEntryNotReady
+
+    hass.data[DOMAIN][entry.entry_id] = vilfo_router
+
+    for platform in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, platform)
+        )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, platform)
+                for platform in PLATFORMS
+            ]
+        )
+    )
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+class VilfoRouterData:
+    """Define an object to hold sensor data."""
+
+    def __init__(self, host, access_token):
+        """Initialize."""
+        self._vilfo = VilfoClient(host, access_token)
+        self.host = host
+        self.available = False
+        self.firmware_version = None
+        self.mac_address = self._vilfo.mac
+        self.data = {}
+        self._unavailable_logged = False
+
+    @property
+    def unique_id(self):
+        """Get the unique_id for the Vilfo Router."""
+        if self.mac_address:
+            return self.mac_address
+
+        if self.host == ATTR_DEFAULT_HOST:
+            return self.host
+
+        return self.host
+
+    @Throttle(DEFAULT_SCAN_INTERVAL)
+    async def async_update(self):
+        """Update data using calls to VilfoClient library."""
+        try:
+            board_information = self._vilfo.get_board_information()
+            self.firmware_version = board_information["version"]
+            boot_time = dt_util.parse_datetime(board_information["bootTime"])
+            uptime = dt_util.now() - boot_time
+            uptime_seconds = round(uptime.total_seconds(), 0)
+            uptime_minutes = round(uptime_seconds / 60, 0)
```

Comment:
I didn't know about the `timestamp` device class, which seems to come in handy here and make it clearer, so I'll change to that :+1:

Perhaps I should rename it from `uptime` to `boot_time` or something as well? I guess the `timestamp` device class is to mark a point in time, rather than a duration? 🕥

Reply:
When I went about to refactor this, I noticed that these relative times are never computed in the sensors themselves, but only during update (which is throttled). I think this should basically be the same as if the API library was doing the calculation. Not sure if it makes any difference in this case? Maybe it still should be avoided to keep things uniform 🙂

Reply:
Yes, please use the timestamp device class and absolute time as state.

Reply:
The lovelace frontend is able to show the absolute timestamp as a relative time.

https://www.home-assistant.io/lovelace/entities/#format